### PR TITLE
Feature: Patient Encounters

### LINKED ISSUE #3074: Proposal: Rename patient "visits" to "encounters"
Apparently the rest of the industry uses "encounters" in the way we use "visits".  Encounters can be anything - from a phone call to a hospitalization.  If we switch from thinking of "Patient Visits" to "Patient Encounters" it will help with user onboarding and provide professional language for our users.

### PR DESCRIPTION
**Pull request by jniles**
This PR addresses the feature request in #3012.

**Design**
The overall design is to make patient admission and discharge much easier to do for the receptionist.  In 1.x, we previously used a separate area on the registration form, but this caused confusion and was limited to looking up patients by their name or BHIMA identifier.  Normally, returning patients provide their hospital number, so this implementation allows a patient to be discovered based on their hospital number.

We also distinguish between "Ambulatory" and "Hospitalized" patients.  The hospitalized patients will need to be discharged, whilst the ambulatory patients could automatically be discharged immediately after admission.  The rationale is that it would be difficult to know if a patient has left the hospital or not.  We may change this behavior with further testing.

Closes #3074.

### CODE REVIEW

**Comment by jniles** [comment]
This should only be merged after the https://github.com/IMA-WorldHealth/bhima-2.X/issues/3408 and https://github.com/IMA-WorldHealth/bhima-2.X/issues/3407 are implemented.

**Comment by mbayopanda** [comment]
@jniles

File: client/src/i18n/en/form.json [hunk 1]
```diff
@@ -512,6 +513,7 @@
             "ORDER_BY": "Order By",
             "PARENT_ACCOUNT": "Parent Account",
             "PASSIVE": "Passive",
+            "PAVILLION":"Pavillion",
```
Comment (jniles) [inline client/src/i18n/en/form.json]: Let's call this "Ward" instead of "Pavilion".

File: client/src/i18n/en/tree.json [hunk 2]
```diff
@@ -81,6 +81,7 @@
     "PAYROLL_RUB_CONFIGURATION" : "Payroll Rubric Configuration",
     "PAYROLL_RUB_MANAGEMENT" : "Payroll Rubric Management",
     "PHARMACY" : "Pharmacy Management",
+    "PAVILLION" : "Pavillion Management",
```
Comment (jniles) [inline client/src/i18n/en/tree.json]: Let's call this "Ward Management"

File: client/src/js/components/bhPDFPrint.js [hunk 3]
```diff
@@ -142,7 +142,10 @@ function bhPDFPrintController($window, $http, $sce, $timeout, Languages, Notify)
       URL.revokeObjectURL($ctrl.blobFileUrl);
     }
 
-    $window.frames[$ctrl.embeddedContentId].removeEventListener('load', printEmbeddedContent);
+    const frame = $window.frames[$ctrl.embeddedContentId];
+    if (frame) {
```
Comment (jniles) [inline client/src/js/components/bhPDFPrint.js:146]: 💃

File: client/src/modules/patients/record/bh-find-document.component.js [hunk 4]
```diff
@@ -104,10 +104,10 @@ function FindDocumentComponent(Patient, Modal, Document, Notify) {
     let ext;
 
     if (mimetype.indexOf('image') > -1) {
-      ext =
-        (mimetype.indexOf('jpg') > -1 || mimetype.indexOf('jpeg') > -1) ? '.jpg' :
-          (mimetype.indexOf('png') > -1) ? '.png' :
-            (mimetype.indexOf('gif') > -1) ? '.gif' : '';
+      /* eslint-disable no-nested-ternary */
+      ext = (mimetype.indexOf('jpg') > -1 || mimetype.indexOf('jpeg') > -1) ? '.jpg'
```
Comment (jniles) [inline client/src/modules/patients/record/bh-find-document.component.js]: We should probably make this into a `MIMETypeService` that performs this operation.  That way, we can test it easier.

File: client/src/modules/pavillion/modals/createUpdate.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+  init();
+
+  function init() {
+    Service.read().then(services => {
+      vm.services = services;
+    });
+
+    if (!vm.isCreating) {
+      Pavillion.read(uuid).then(pavillion => {
```
Comment (jniles) [inline client/src/modules/pavillion/modals/createUpdate.js]: We should probably have a `.catch()` here.

File: client/src/modules/pavillion/modals/createUpdate.html [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    </div>
+
+    <div class="form-group"
+      ng-class="{'has-error' : ModalForm.$submitted && ModalForm.service.$invalid }">
+      <label class="control-label" translate>FORM.LABELS.SERVICE</label>
+      <select
+        class="form-control"
+        name="service"
+        ng-model="ModalCtrl.pavillion.service_id"
+        ng-options="service.id as service.name for service in ModalCtrl.services"
```
Comment (jniles) [inline client/src/modules/pavillion/modals/createUpdate.html]: Can we use [bhServiceSelect](https://github.com/IMA-WorldHealth/bhima/blob/master/client/src/js/components/bhServiceSelect.js) here?

File: client/src/modules/pavillion/pavillion.html [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    </ol>
+
+    <div class="toolbar">
+      <div class="toolbar-item">
+        <button class="btn btn-default text-capitalize" id="create" ng-click="PavillionCtrl.createPavillion()" data-method="create">
+          <span class="fa fa-plus"></span> <span translate>FORM.LABELS.ADD</span>
+        </button>
+      </div>
+
+      <div class="toolbar-item">
```
Comment (jniles) [inline client/src/modules/pavillion/pavillion.html]: Can we use `bhToggleFilter` here?

File: client/src/modules/pavillion/pavillion.html [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  <div class="container-fluid">
+    <div id="pavillion-grid"
+      ui-grid="PavillionCtrl.gridOptions"
+      class="grid-full-height"
+      ui-grid-auto-resize
+      ui-grid-resize-columns>
+
+      <bh-grid-loading-indicator
+        loading-state="PavillionCtrl.loading"
+        empty-state="PavillionCtrl.gridOptions.data.length === 0">
```
Comment (jniles) [inline client/src/modules/pavillion/pavillion.html]: We should have an "error state" here.

File: server/models/migrations/v0.8.0-v1.0.0/migrate.sql [hunk 9]
```diff
@@ -962,6 +962,13 @@ UPDATE role_unit SET unit_id = 199 WHERE unit_id = 159;
 DELETE FROM unit WHERE id = 159;
 DELETE FROM report WHERE id = 9;
 
+INSERT INTO `account_reference_item`(`id`, `account_reference_id`, `account_id`, `is_exception`, `credit_balance`, `debit_balance`)
+  VALUES (1,1,2958,0,0,0),(5,5,2996,0,0,0),(6,5,3012,0,0,0),(24,20,3282,0,0,0),(27,23,3346,0,0,0),(29,25,3390,0,0,0),(30,25,3382,0,0,0),(31,26,3385,0,0,0),(45,2,2973,0,0,0),(55,22,3351,0,0,0),(56,3,2967,0,0,0),(57,4,2984,0,0,0),(58,36,3292,0,0,0),(59,37,3321,0,0,0),(60,38,3355,0,0,0),(61,39,3385,0,0,0),(62,40,2987,0,0,0),(63,41,3036,0,0,0),(64,6,3058,0,0,0),(65,6,3100,0,0,0),(66,7,3132,0,0,0),(67,8,3158,0,0,0),(68,9,3178,0,0,0),(69,42,3390,0,0,0),(70,10,3234,0,0,0),(71,10,3260,0,0,0),(75,43,3372,0,0,0),(76,24,3407,0,0,0),(77,44,3387,0,0,0),(78,45,3218,0,0,0),(79,46,3277,0,0,0),(80,47,3415,0,0,0),(81,48,3425,0,0,0),(82,48,3436,0,0,0),(83,48,3442,0,0,0),(84,49,3411,0,0,0),(85,11,3419,0,0,0),(86,11,3431,0,0,0),(87,12,3447,0,0,0);
```
Comment (jniles) [inline server/models/migrations/v0.8.0-v1.0.0/migrate.sql:966]: Let's move these into `next/migrate.sql`.

File: server/models/schema.sql [hunk 10]
```diff
@@ -1667,6 +1668,17 @@ CREATE TABLE `service` (
   FOREIGN KEY (`profit_center_id`) REFERENCES `profit_center` (`id`) ON UPDATE CASCADE
 ) ENGINE=InnoDB  DEFAULT CHARACTER SET = utf8mb4 DEFAULT COLLATE = utf8mb4_unicode_ci;
 
+DROP TABLE IF EXISTS `pavillion`;
```
Comment (jniles) [inline server/models/schema.sql]: We need to put this in `next/migrate.sql`.

**Review by jniles — COMMENTED** [review]
@mbayopanda great work on this.  I've made a few comments to clean up the code.

**Comment by mbayopanda** [comment]
@jniles

**Review by jniles — COMMENTED** [review]
LGTM!  Excellent work.

bors r+

**Pull request merged**